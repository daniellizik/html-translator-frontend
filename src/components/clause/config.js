export const QUERY = {
  targets: {
    NODE_NAME: {
      rules: ['EQUALS','NOT_EQUALS','LIKE','NOT_LIKE']
    },
    TEXT: {
      rules: ['EQUALS','NOT_EQUALS','LIKE','NOT_LIKE','REGEX']
    },
    ATTR_KEY: {
      rules: ['EQUALS','NOT_EQUALS','LIKE','NOT_LIKE','REGEX']
    },
    ATTR_VAL: {
      rules: ['EQUALS','NOT_EQUALS','LIKE','NOT_LIKE','REGEX']
    }
  }
}

export const MUTATION = {
  targets: {
    NODE_NAME: {
      rules: ['ALL_REPLACE']
    },
    TEXT: {
      rules: ['START_OF','END_OF','ALL_REPLACE','REGEX_REPLACE']
    },
    ATTR_KEY: {
      rules: ['START_OF','END_OF','REGEX_REPLACE','ALL_REPLACE','CONCAT']
    },
    ATTR_VAL: {
      rules: ['START_OF','END_OF','REGEX_REPLACE','ALL_REPLACE','CONCAT']
    }
  }
}

export const defaultQuery = {
  type: 'QUERY',
  target: 'TEXT',
  targetValue: '',
  rule: 'EQUALS',
  ruleValue: '',
  ruleValueFlags: ''
}

export const defaultMutation = {
  type: 'MUTATION',
  target: 'TEXT',
  targetValue: '',
  rule: 'START_OF',
  ruleValue: '',
  ruleValueFlags: ''
}

export const defaultClause = {
  name: '',
  minimized: false,
  active: false,
  rules: [defaultQuery],
  view: []
}