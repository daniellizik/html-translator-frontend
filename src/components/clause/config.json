{
  "query": {
    "targets": {
      "nodeName": {
        "rules": [
          "equals",
          "notEquals",
          "like",
          "notLike"
        ]
      },
      "text": {
        "rules": [
          "equals",
          "notEquals",
          "like",
          "notLike",
          "regex"
        ]
      },
      "attrKey": {
        "rules": [
          "equals",
          "notEquals",
          "like",
          "notLike",
          "regex"
        ]
      },
      "attrVal": {
        "rules": [
          "equals",
          "notEquals",
          "like",
          "notLike",
          "regex"
        ]
      }
    },
    "defaultQuery": {
      "type": "query",
      "target": "text",
      "targetValue": "",
      "rule": "like",
      "ruleValue": "",
      "ruleValueFlags": ""
    }
  },

  "mutator": {
    "targets": {
      "nodeName": {
        "rules": [
          "allReplace"
        ]
      },
      "text": {
        "rules": [
          "startOf",
          "endOf",
          "allReplace",
          "regexReplace"
        ]
      },
      "attrKey": {
        "rules": [
          "startOf",
          "endOf",
          "regexReplace",
          "allReplace",
          "add"
        ]
      },
      "attrVal": {
        "rules": [
          "startOf",
          "endOf",
          "regexReplace",
          "allReplace",
          "add"
        ]
      }
    },
    "defaultMutation": {
      "type": "mutation",
      "target": "text",
      "targetValue": "",
      "rule": "startOf",
      "ruleValue": "",
      "ruleValueFlags": ""
    }
  }

}


