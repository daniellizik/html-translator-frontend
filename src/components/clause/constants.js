export const CLAUSE_ADD = '@CLAUSE/CLAUSE_ADD'
export const CLAUSE_REMOVE = '@CLAUSE/CLAUSE_REMOVE'
export const CLAUSE_CHANGE_NAME = '@CLAUSE/CHANGE_NAME'

export const CLAUSE_INVALID_REGEXP = '@CLAUSE/CLAUSE_INVALID_REGEXP'
export const CLAUSE_INVALID_ATTRKEY = '@CLAUSE/CLAUSE_INVALID_ATTRKEY'

export const QUERY_ADD = '@CLAUSE/QUERY_ADD'
export const QUERY_REMOVE = '@CLAUSE/QUERY_REMOVE'

export const QUERY_CHANGE_TARGET = '@CLAUSE/QUERY_CHANGE_TARGET'
export const QUERY_CHANGE_TARGET_VALUE = '@CLAUSE/QUERY_CHANGE_TARGET_VALUE'
export const QUERY_CHANGE_RULE = '@CLAUSE/QUERY_CHANGE_RULE'
export const QUERY_CHANGE_RULE_VALUE = '@CLAUSE/QUERY_CHANGE_RULE_VALUE'
export const QUERY_CHANGE_RULE_VALUE_FLAGS = '@CLAUSE/QUERY_CHANGE_RULE_VALUE_FLAGS'

export const MUTATION_ADD = '@CLAUSE/MUTATION_ADD'
export const MUTATION_REMOVE = '@CLAUSE/MUTATION_REMOVE'

export const MUTATION_ADD_ATTR = '@CLAUSE/MUTATION_ADD_ATTR'
export const MUTATION_REMOVE_ATTR_BY_KEY = '@CLAUSE/MUTATION_REMOVE_ATTR_BY_KEY'
export const MUTATION_REMOVE_ATTR_BY_VALUE = '@CLAUSE/MUTATION_REMOVE_ATTR_BY_VALUE'
export const MUTATION_REMOVE_ALL_ATTRS = '@CLAUSE/MUTATION_REMOVE_ALL_ATTRS'

