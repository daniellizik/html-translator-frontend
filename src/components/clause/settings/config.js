export const targets = ['NODE_NAME', 'TEXT', 'ATTR_KEY', 'ATTR_VAL']
export const behaviors = ['APPLY_TO_CHILDREN', 'APPLY_TO_NODE']

export const QUERY = {
  targets: {
    NODE_NAME: {
      rules: [
        'EQUALS', 
        'NOT_EQUALS', 
        'LIKE', 
        'NOT_LIKE',
        'REGEX'
      ],
      behaviors: []
    },
    TEXT: {
      rules: [
        'EQUALS', 
        'NOT_EQUALS', 
        'LIKE', 
        'NOT_LIKE',
        'REGEX'
      ],
      behaviors: []
    },
    ATTR_KEY: {
      rules: [
        'EQUALS', 
        'NOT_EQUALS', 
        'LIKE', 
        'NOT_LIKE',
        'REGEX',
        'HAS_NONE',
        'HAS_AT_LEAST'
      ],
      behaviors: []
    },
    ATTR_VAL: {
      rules: [
        'EQUALS', 
        'NOT_EQUALS', 
        'LIKE', 
        'NOT_LIKE',
        'REGEX'
      ],
      behaviors: []
    }
  }
}

export const MUTATION = {
  targets: {
    NODE_NAME: {
      rules: ['ALL_REPLACE', 'START_OF','END_OF'],
      behaviors
    },
    TEXT: {
      rules: ['ALL_REPLACE', 'START_OF', 'END_OF', 'REGEX_REPLACE'],
      behaviors
    },
    ATTR_KEY: {
      rules: ['ALL_REPLACE', 'START_OF', 'END_OF', 'REGEX_REPLACE'],
      behaviors
    },
    ATTR_VAL: {
      rules: ['ALL_REPLACE', 'START_OF', 'END_OF', 'REGEX_REPLACE'],
      behaviors
    }
  }
}

// hardcoded for parse5 ast spec
export const targetMap = {
  TEXT: 'value',
  NODE_NAME: 'nodeName'
}

export const defaultQuery = {
  targetValue: '',
  rule: 'EQUALS',
  ruleValue: '',
  ruleValueFlags: ''
}

export const defaultMutation = {
  targetValue: '',
  rule: 'START_OF',
  ruleValue: '',
  ruleValueFlags: '',
  behavior: ''
}

export const defaultClause = {
  name: '',
  minimized: false,
  active: false,
  target: 'TEXT',
  type: 'CLAUSE',
  queries: [defaultQuery],
  mutations: [],
  view: []
}