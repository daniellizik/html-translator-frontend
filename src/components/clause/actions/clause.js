import * as constants from '../constants'

export default {
  activate: (clauseIndex) => ({
    type: constants.CLAUSE_ACTIVATE,
    clauseIndex
  }),
  add: () => ({
    type: constants.CLAUSE_ADD
  }),
  remove: (clauseIndex) => ({
    type: constants.CLAUSE_REMOVE,
    clauseIndex
  }),
  removeAll: () => ({
    type: constants.CLAUSE_REMOVE_ALL
  }),
  changeName: (clauseIndex, name) => ({
    type: constants.CLAUSE_CHANGE_NAME,
    clauseIndex,
    name
  }),
  changeTarget: (target, clauseIndex) => ({
    type: constants.CLAUSE_CHANGE_TARGET,
    target,
    clauseIndex
  }),
  setCurrentMutation: (currentMutation) => ({
    type: constants.CLAUSE_VIEW_MUTATIONS,
    currentMutation
  })
}