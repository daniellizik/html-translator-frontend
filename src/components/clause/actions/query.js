import * as constants from '../constants'

export const remove = (clauseIndex, ruleIndex) => ({
  type: constants.QUERY_REMOVE,
  clauseIndex,
  ruleIndex,
  ruleType: 'QUERY'
})

export const add = (clauseIndex) => ({
  type: constants.QUERY_ADD,
  clauseIndex
})

// for async search
export const changeInit = () => ({ type: constants.QUERY_CHANGE_INIT })

export const changeError = (error) => ({
  type: constants.QUERY_CHANGE_ERROR,
  error
})

export const changeDone = (props) => ({
  ...props,
  type: constants.QUERY_CHANGE_DONE
})

export const changeTargetValue = (targetValue, clauseIndex, ruleIndex) => ({
  type: constants.QUERY_CHANGE_TARGET_VALUE,
  targetValue,
  clauseIndex,
  ruleIndex
})

export const changeTargetValueAsync = (tv, ci, r) => (dispatch) => {
  
}

export const changeRule = (rule, clauseIndex, ruleIndex) => ({
  type: constants.QUERY_CHANGE_RULE,
  rule,
  clauseIndex,
  ruleIndex
})

export const changeTarget = (target, clauseIndex, ruleIndex) => ({
  type: constants.QUERY_CHANGE_TARGET,
  target,
  clauseIndex,
  ruleIndex
})

export const changeRuleValue = (ruleValue, clauseIndex, ruleIndex) => ({
  type: constants.QUERY_CHANGE_RULE_VALUE,
  ruleValue,
  clauseIndex,
  ruleIndex
})

export const changeRuleValueFlags = (ruleValueFlags, clauseIndex, ruleIndex) => ({
  type: constants.QUERY_CHANGE_RULE_VALUE_FLAGS,
  ruleValueFlags,
  clauseIndex,
  ruleIndex
})
