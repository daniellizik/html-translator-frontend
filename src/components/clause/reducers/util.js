import * as constants from '../constants'
import * as targets from '../settings/targets'
import * as rules from '../settings/rules'
import * as behaviors from '../settings/behaviors'
import { targetMap } from '../settings/config'

export const errorHandler = ({message}) => {
  return message
}

export const mutateList = (list = [], view = []) => {
  return list.reduce((acc, node) => { 
    const index = view.indexOf(node.id)
    return index < 0 
      ? [...acc, node]
      : [...acc, {...node, view: true}]
  },[])
}

// take view, mutations and apply them to FULL LIST, not open
// this makes it easier for xml tree to consume (it just spits out entire list)
export const mutationDenormalizer = (clause, list = []) => {
  const { target, view, mutations } = clause
  // mutations might have to mutate entire list, so give a copy to keep it immutable
  return list.slice().reduce((acc, node, i, arr) => {
    // only mutate items in view
    return view.indexOf(node.id) < 0 ? [...acc, node] : [
      ...acc,
      // this needs to account for mutation behavior
      mutations.reduce((mutatedNode, mutation) => {
        const targetProp = targetMap[target]
        const params = {...mutation, before: mutatedNode[targetProp]}
        const ruleResult = rules[mutation.rule](params)
        // todo: implement this
        // const behaviorResult = behaviors[mutation.behavior](mutatedNode, mutation, arr, ruleResult)
        return {
          ...mutatedNode,
          [targetProp]: ruleResult
        }
      }, {...node})
    ] 
  }, [])
}

export const reduceRuleProp = (ruleType, state, action, prop) => {
  return {
    ...state,
    clauses: state.clauses.map((c, i) => (
      i !== action.clauseIndex ? c : {
        ...c,
        [ruleType]: c[ruleType].map((r, j) => (
          j !== action.ruleIndex ? r : {
            ...r,
            [prop]: action[prop]
          }
        ))
      }
    ))
  }
}

export const reduceView = (clauseIndex, {list}, clauses) => {
  return clauses.reduce((acc, clause, index) => {
    if (index !== clauseIndex)
      return [...acc, clause]
    const view = list.open.reduce((ids, node) => {
      const clauseResult = clause.queries.reduce((bool, obj) => {
        const targetResult = targets.query[clause.target](node, {...obj, rule: rules[obj.rule]})
        if (targetResult === false || bool === false)
          return false
        else if (targetResult === true)
          return true
      }, null)
      return !clauseResult ? ids : [...ids, node.id]
    }, [])
    return [
      ...acc,
      { ...clause, view }
    ]
  }, [])
}

export const reduceClauses = (state, action, key) => {
  let nextState
  const clauses = state.clauses.map((clause, clauseIndex) => {
    return clauseIndex !== action.clauseIndex ? clause : {
      ...clause,
      queries: clause.queries.map((query, ruleIndex) => {
        return ruleIndex !== action.ruleIndex ? query : { ...query, [key]: action[key] }
      })
    }
  })
  try {
    nextState = {
      ...state,
      clauses: reduceView(action.clauseIndex, state.slave, clauses)
    }
  }
  catch(e) {
    process.env.NODE_ENV === 'development' && console.warn('error', e)
    return {
      ...state,
      clauses,
      error: errorHandler(e)
    }
  }
  return nextState
}