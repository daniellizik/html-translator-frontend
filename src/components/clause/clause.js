import React, { Component, PropTypes } from 'react'
import { connect } from 'react-redux'
import { bindActionCreators } from 'redux'
import policyValidator from '~/src/components/clause/policies/validator'
import { queryActions, mutateActions, clauseActions } from '~/src/components/clause/actions/index'
import * as config from '~/src/components/clause/settings/config'
import { ChangeTarget } from './clauses'
import { ToolTip, ChangeTargetExplanation } from '~/src/components/explanation'

const mapDispatchToProps = (dispatch) => ({
  queryActions: bindActionCreators(queryActions, dispatch),
  mutateActions: bindActionCreators(mutateActions, dispatch),
  clauseActions: bindActionCreators(clauseActions, dispatch)
})

const mapStateToProps = (state) => ({
  ...state,
  currentMutation: state.slave.currentMutation,
  onboardingStep: state.onboarding.step
})

export const Clause = connect(mapStateToProps, mapDispatchToProps)((props) => (
  <div class="py-0 bg-lightGrey">
    {policyValidator(props)}
    <div class="row py-2 m-0">
      <div class="col-4">
        <span onClick={() => props.actionSet[props.type].remove(props.clauseIndex, props.ruleIndex)}>
          {({QUERY: 'remove this query', MUTATION: 'remove this mutation'})[props.type]}
        </span>
      </div>
    </div>
  </div>
))

export const MaximizedClause = connect(mapStateToProps, mapDispatchToProps)(({ 
  currentMutation, 
  clauseActions, 
  queryActions, 
  mutateActions, 
  clauseIndex, 
  clauseGroup,
  onboardingStep
}) => (
  <div
    key={clauseIndex} 
    class="col-12 mx-0 mb-3 py-3 c-white bg-middleGrey">
    <div class="row px-3">
      <label class="col-6 mb-2 mx-0 pl-0 pr-2">
        <p>change target</p>
        <ToolTip
          placement="right"
          destroyTooltipOnHide={true}
          visible={onboardingStep === 4}
          trigger={clauseIndex === 0 ? ['hover'] : []}
          overlay={<ChangeTargetExplanation />}>
          <select 
            class="form-control custom-select"
            onFocus={() => clauseActions.activate(clauseIndex)}
            onChange={({target}) => clauseActions.changeTarget(target.value, clauseIndex)}>
            {config.targets.map((p, j) => (
              <option value={p} key={j}>
                {p}
              </option>
            ))}
          </select>
        </ToolTip>
      </label>

      <div class="col-12 mb-2 mx-0 px-0">
        <button class="btn mr-2 my-1" onClick={() => clauseActions.remove(clauseIndex)}>
          remove this clause
        </button>
        <button class="btn mr-2 my-1" onClick={() => clauseActions.denormalize(currentMutation === clauseIndex ? -1 : clauseIndex)}>
          {currentMutation === clauseIndex ? 'hide mutations' : 'view mutations'}
        </button>
        <button class="btn mr-2 my-1" onClick={() => queryActions.add(clauseIndex)}>
          add a query
        </button>
        <button class="btn mr-2 my-1" onClick={() => mutateActions.add(clauseIndex)}>
          add a mutation
        </button>
      </div>
      <div class="col-12 m-0 p-0">
        <div class="row m-0 p-0">
          <div class="col m-0 p-0">queries</div>
        </div>
        {clauseGroup.queries.map((clause, ruleIndex, {length}) => (
          <Clause 
            type="QUERY"
            actionSet={queryActions}
            clause={clause} 
            isLast={ruleIndex === length - 1} 
            clauseIndex={clauseIndex} 
            ruleIndex={ruleIndex} 
            key={`q-${clauseIndex}-${ruleIndex}`} />
        ))}
        <div class="row m-0 p-0">
          <div class="col m-0 p-0">mutations</div>
        </div>
        {clauseGroup.mutations.map((clause, ruleIndex, {length}) => (
          <Clause 
            type="MUTATION"
            actionSet={mutateActions}
            clause={clause} 
            isLast={ruleIndex === length - 1} 
            clauseIndex={clauseIndex} 
            ruleIndex={ruleIndex} 
            key={`m-${clauseIndex}-${ruleIndex}`} />
        ))}
      </div>
    </div>
  </div>
))

export const MinifiedClause = connect(mapStateToProps, mapDispatchToProps)((props) => (
  null
))