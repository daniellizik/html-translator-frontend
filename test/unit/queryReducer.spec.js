import * as queries from '~/test/storyFixtures/query'
import * as constants from '~/src/components/clause/constants'
import { defaultQuery } from '~/src/components/clause/settings/config'
import { filterText } from '~/src/util'

describe('query reducer', () => {

  Object.keys(queries).forEach(q => {
    it('should set clauses', () => {
      expect(q[q.length - 1].clauses).toMatchSnapshot()
    })
  })

})