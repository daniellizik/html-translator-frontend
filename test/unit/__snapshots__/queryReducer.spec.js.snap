exports[`query reducer should set clauses 1`] = `undefined`;

exports[`query reducer should set clauses 2`] = `undefined`;

exports[`query reducer should set clauses 3`] = `undefined`;

exports[`query reducer should set clauses 4`] = `undefined`;

exports[`query reducer should set clauses 5`] = `undefined`;

exports[`query reducer should set clauses 6`] = `undefined`;

exports[`query reducer should set clauses 7`] = `undefined`;

exports[`query reducer should set clauses 8`] = `undefined`;

exports[`query reducer should set clauses 9`] = `undefined`;
